bundle/edu.gemini.spModel.core/src/main/scala/edu/gemini/spModel/core/ProperMotion.scala

```diff
@@ -11,16 +15,126 @@ import Scalaz._
+  private val radsPerAsec: Double = Angle.fromArcsecs(1).toRadians
+  private val TwoPi:       Double = 6.283185307179586476925286766559
+  private val Million:     Double = 1000000.0
+
+  private type Vec2 = (Double, Double)
+  private type Vec3 = (Double, Double, Double)
+
+  // Scalar multiplication and addition for Vec3.
+  private implicit class Vec3Ops(a: Vec3) {
+    def *(d: Double): Vec3 =
+      (a._1 * d, a._2 * d, a._3 * d)
+
+    def |+|(d: Double): Vec3 =
+      (a._1 + d, a._2 + d, a._3 + d)
+
+    def |+|(other: Vec3): Vec3 =
+      (a._1 + other._1, a._2 + other._2, a._3 + other._3)
+  }
+
+  /**
+   * Calculator to determine the new coordinates due to proper motion from epoch after elapsed years.
+   * @param baseCoordinates base coordinates of the target
+   * @param daysPerYear     length of epoch year in fractional days
+   * @param properVelocity  proper velocity per epoch year
+   * @param radialVelocity  radial velocity (km / s, positive if receding)
+   * @param parallax        parallax
+   * @param elapsedYears    elapsed time in epoch years
+   * @return
+   */
+  def properMotionCalculator(baseCoordinates: Coordinates,
+                             daysPerYear: Double,
+                             properVelocity: Offset,
+                             radialVelocity: Double,
+                             parallax: Double,
+                             elapsedYears: Double): Coordinates = {
+    // We want the baseCoordinates in radians.
+    val (ra,  dec)  = (baseCoordinates.ra.toAngle.toRadians, baseCoordinates.dec.toAngle.toRadians)
+    val (dRa, dDec) = (properVelocity.p.toAngle.toRadians,   properVelocity.q.toAngle.toRadians)
+
+    val pos: Vec3 = {
+      val cd = cos(dec)
+      (cos(ra) * cd, sin(ra) * cd, sin(dec))
+    }
+
+    // Change per year due to radial velocity and parallax. The units work out to asec/y.
+    val dPos1: Vec3 =
+        pos            *
+        daysPerYear    *
+        secsPerDay     *
+        radsPerAsec    *
+        auPerKm        *
+        radialVelocity *
+        parallax
+
+    // Change per year due to proper velocity
+    val dPos2 = (
+      -dRa * pos._2 - dDec * cos(ra) * sin(dec),
+       dRa * pos._1 - dDec * sin(ra) * sin(dec),
+                      dDec *           cos(dec)
+    )
+
+    // Our new position (still in polar coordinates). `|+|` here is scalar addition.
+    val pp = pos |+| ((dPos1 |+| dPos2) * elapsedYears)
+
+    // Back to spherical
+    val (x, y, z) = pp
+    val r    = hypot(x, y)
+    val rap  = if (r === 0.0) 0.0 else atan2(y, x)
+    val decp = if (z === 0.0) 0.0 else atan2(z, r)
+    val rapp = {
+      // Normalize to [0 .. 2π)
+      val rem = rap % TwoPi
+      if (rem < 0.0) rem + TwoPi else rem
+    }
+
+    // TODO: What to do if Declination is None?
+    Coordinates(RA.fromAngle(Angle.fromRadians(rapp)),
+                Declination.fromAngle(Angle.fromRadians(decp)).getOrElse(Declination.zero))
```

Comment:
We should probably fail if `Declination.fromAngle` is `None` since it implies an error in our logic.  So maybe

```
Declination.fromAngle(...).getOrElse(sys.error(s"Invalid declination: $decp radians"))
```

or something.